### ISSUE TITLE
Unwanted Sigmoid in BinaryROC/BinaryPrecisionRecallCurve Causes Incorrect Behavior in Batch-wise Updates

### ISSUE THREAD

**Issue by PussyCat0700** [Lightning-AI/torchmetrics#3179]
❗ Problem Summary

When using BinaryROC or BinaryPrecisionRecallCurve in a typical PyTorch training loop, it's common to call `.update()` multiple times with batch-wise logits and then `.compute()` at the end. However, torchmetrics applies an implicit sigmoid transformation to predictions inside `_binary_precision_recall_curve_format()`.

This breaks the fundamental assumption of .update() being safe for incremental accumulation, because:

- Applying Sigmoid batch-wise does not preserve global ordering of scores.

- Metrics like ROC/EER/DET depend solely on relative score ordering, not absolute values.

This leads to inconsistent or incorrect metrics, especially for EER/DET.

📍 Code Reference
https://github.com/Lightning-AI/torchmetrics/blob/e1679aeeb81e8605a07c0c6e79df12f0d88294a9/src/torchmetrics/classification/precision_recall_curve.py#L166

calls:

https://github.com/Lightning-AI/torchmetrics/blob/e1679aeeb81e8605a07c0c6e79df12f0d88294a9/src/torchmetrics/functional/classification/precision_recall_curve.py#L185

which compels sigmoid on any batch with logits outside [0, 1] in:

https://github.com/Lightning-AI/torchmetrics/blob/e1679aeeb81e8605a07c0c6e79df12f0d88294a9/src/torchmetrics/utilities/compute.py#L231-L243

🔬 Why This is Harmful

Let’s assume I’m doing:

```
for embs, labels in dataloader:
    scores = cosine_similarity(embs[0], embs[1])
    metric.update(logits, scores)  # logits, not sigmoid(logits)
```

- Expected: the metric accumulates raw scores and uses their global ranking that can be used to compute ROC/EER.

- Reality: each batch's logits are likely to be nonlinearly transformed, so relative rankings across batches are broken.

This can result in wrong DET/EER curve with the FPR and TPR provided like this, with inconsistent results depending on batch size or number of updates.

✅ Suggested Fixes

As this sigmoid function is not really necessary despite the presence of logits outside [0,1] in some batches (mostly common in cosine similarities), we suggest to:

- Remove internal sigmoid and document that users should pass probabilities if needed

- Or, expose a flag like apply_sigmoid=True/False in all relevant classes (BinaryROC, BinaryPrecisionRecallCurve, etc.)

- Or, raise warning or error if predictions are outside [0,1] range

🙏 Final Words
We understand that defaulting to sigmoid may feel "user-friendly" for those passing logits by mistake. But for any serious usage involving:

- metrics tracking during training

- batch-level updates

- scientific reproducibility

this behavior is silently harmful and counterintuitive.

We hope to see `.update()` remain safe and composable across time, without hidden nonlinearities.

Thank you for your great work on torchmetrics and for considering this issue!

**Comment by PussyCat0700** [comment]
> Nice finding, could you send a draft PR? 🐰

Sure, I opened a draft PR in https://github.com/Lightning-AI/torchmetrics/issues/3179#issuecomment-3069139964
